### PR TITLE
Vindex for helping in migrations to vitess

### PR DESCRIPTION
**Pull request by chrisgillis**
The NumericStaticMap VIndex will allow you to provide a JSON file of alternative mappings. Behaves the same as Numeric with the exception of checking the map first.

More context on the mailing list here: https://groups.google.com/forum/#!msg/vitess/-hPmDJj8x1M/RVlrSqf-AAAJ

---
This change is [Reviewable](https://reviewable.io/reviews/youtube/vitess/1924)

### CODE REVIEW

**Comment by chrisgillis** [comment]
Had to add an e-mail address to my github profile. CLA should be OK now.

**Comment by chrisgillis** [comment]
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 27 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOgcyo0iIGneeAD_5jk:-KOgolkRIyHCHBPhdbE6:b1x78yv) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L27)):*
<details><summary><i>Previously, enisoc-bot (Anthony Yeh) wrote…</i></summary>
> It may be tricky to distribute this file along with every server that runs vtgate, vtctld, vtworker, etc. Also if the file is changed, there is the danger of the servers getting out of sync.
> 
> If the static map is expected to be reasonably small, could we just store the mapping itself in a vindex param value (perhaps some kind of simple string syntax rather than full JSON)? Then it would be stored in topology along with the rest of vschema.
</details>
Thoughts:

Our primary use case for this change is migration so it's unlikely to ever change for us at least. We bring our app down for maintenance and generate the JSON file, distribute it and we're done.

Our shard keys are so noncontiguous  that our json file with the needed mappings was 1.1MB. Eliminating json syntax chars would probably cut the filesize by 25% or so at the cost of being slightly less intuitive to work with. I think adding to topo was Sugo's suggestion for a later date after the completion of something else. Details are in the bottom of the mailing list thread linked in the PR.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 32 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOgdf_vkoTZEce2tmsX:-KOgofmdSPtnM5LuFTqj:b-g00w5k) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L32)):*
<details><summary><i>Previously, enisoc-bot (Anthony Yeh) wrote…</i></summary>
> The conventional name for an error variable is `err`. The name `error` is actually a pre-defined interface that you're shadowing here.
</details>
Copied style from numeric.go, will refactor though

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 72 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOgehEmj9MUxe9u2ASo:-KOgnvskQEykyuZlpsgp:b15imhd) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L72)):*
<details><summary><i>Previously, enisoc-bot (Anthony Yeh) wrote…</i></summary>
> This looks like you would need an individual map entry for every existing account ID. I thought the idea was to map ID ranges to keyspace ID ranges?
> 
> Something like:
> 
> 1-5 => 0-4
> 11-15 => 5-9
> 
> Is that not the plan anymore?
</details>
I think we forgot about the ranges plan. It was very easy to create a file for the individual maps. I can refactor this if you think it is necessary.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 81 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOgfvpMnfEcOldjH0_7:-KOgnXthjF4mvmHckqrv:bbtuzqx) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L81)):*
<details><summary><i>Previously, enisoc-bot (Anthony Yeh) wrote…</i></summary>
> You shouldn't need to do the `[]byte(...)` cast since `keybytes[:]` should already be of type byte slice.
</details>
I can change this. I copied from numeric.go, see: https://github.com/youtube/vitess/blob/master/go/vt/vtgate/vindexes/numeric.go

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1924)*

**Comment by chrisgillis** [comment]
Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 17 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOh7g4v_RH_IF7XNdls:-KOhMCBMcYNHfU3_HzyM:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L17)):*
<details><summary><i>Previously, sougou-bot wrote…</i></summary>
> This should preferably be map[int64]uint64, or even map[uint64]uint64.
</details>
Done.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 27 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOgcyo0iIGneeAD_5jk:-KOhEqTZ1PyvZVpSF7e-:br8q2xm) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L27)):*
<details><summary><i>Previously, sougou-bot wrote…</i></summary>
> The file based approach does make this non-generic because there's no good story for resharding etc. But I don't have a better suggestion right now. The only other option I can think of is to store the info in a db table.
</details>
Splitting shards should still work at least.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 29 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOgdLhsFGOJKDgKtGOE:-KOhJM8OX7UH0pMtaN_s:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L29)):*
<details><summary><i>Previously, enisoc-bot (Anthony Yeh) wrote…</i></summary>
> This should just return an error instead of panicking. Otherwise you could crash the server by loading invalid vschema.
</details>
Done.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 32 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOgdf_vkoTZEce2tmsX:-KOhEQobhbznYXFX9gPR:bn0tsgq) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L32)):*
<details><summary><i>Previously, sougou-bot wrote…</i></summary>
> I only see err used as var name in numeric.go. Can you show me where you saw this construct?
</details>
Oh whoops, I apologize, I was looking at the return type and mistook it for a named return type. Golang beginner here :)

Done.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 72 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOgehEmj9MUxe9u2ASo:-KOhFZqlW9AdKWUd6f-o:b-n990yw) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L72)):*
<details><summary><i>Previously, sougou-bot wrote…</i></summary>
> A simple map won't work if we went with ranges, right? But I think the range thing may be more flexible, and will scale better with number of values.
</details>
Yeah, the data structure would be a little bit more complex to handle ranges and the lookup would be a little more costly as well. For my particular use case of migrating from scalebase, our ranges are so noncontiguous I don't know that it would make much of a difference to use ranges. It may be a little awkward to parse out as well.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 81 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOgfvpMnfEcOldjH0_7:-KOhJnVAFaDtOmrayRYg:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L81)):*
<details><summary><i>Previously, sougou-bot wrote…</i></summary>
> Agree with the suggested change. Can you also change numeric.go accordingly?
</details>
Done.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 113 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOgeJ32sUiy-duoOErI:-KOhKStBhWxC6rtV05BV:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L113)):*
<details><summary><i>Previously, enisoc-bot (Anthony Yeh) wrote…</i></summary>
> If `ParseInt` returns an error, we should abort and return it to our caller.
</details>
Done.

---

*[go/vt/vtgate/vindexes/numeric_static_map_test.json, line 1 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOhACKfBOgzZns9uCbW:-KOhOrPieo_yivCPpYQO:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map_test.json#L1)):*
<details><summary><i>Previously, sougou-bot wrote…</i></summary>
> Can you move this file to data/test/vtgate?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1924)*

**Comment by chrisgillis** [comment]
Review status: 3 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 74 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOhTiCJOtV9XQWE7M_R:-KOhXVETO7Dxd3H4W0o0:bkvfyk6) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L74)):*
<details><summary><i>Previously, enisoc-bot (Anthony Yeh) wrote…</i></summary>
> It seems like this is not reversible in the general case, since the mapping x->y would mean both x and y map to y. Is there something about your generated mappings that makes this reversible? (e.g. if all mappings are actually x<->y swaps)
</details>
I'm not certain I fully understand context and question. We're mapping each shardkey to a unique keyspace id if and only if the generated keyspace id is different than the shardkey AND the shardkey doesnt fall into the keyspace id range it needs to. I'd imagine if I can go from shardkey: 10 to keyspaceid: 4, I should be able to go from 4 to 10 as well given the uniqueness of both sides of the map. Not sure if this answers your question :(

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 76 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOh69Lpd7oDbiffP8Jj:-KOhWZ3wiuxoHnrEhmK0:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L76)):*
<details><summary><i>Previously, enisoc-bot (Anthony Yeh) wrote…</i></summary>
> This err check should be immediately after the place the err variable is set, so right under `getNumber`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1924)*

**Comment by chrisgillis** [comment]
Review status: 3 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 92 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOh7Pd98YdQnlCf7gBe:-KOl3OixqCGcLKtwl1u3:b-ioddsl) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L92)):*
<details><summary><i>Previously, enisoc-bot (Anthony Yeh) wrote…</i></summary>
> @sougou: What's the implication if `ReverseMap` is slow? Should this be done with a pre-computed reverse map generated while loading the table?
</details>
This seemed like a good idea anyway as mem cost is probably minimal in most use cases. 

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1924)*

**Comment by chrisgillis** [comment]
Review status: 3 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 92 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOh7Pd98YdQnlCf7gBe:-KOlA_Ft5hl_eGYH9jpL:bxotfbr) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L92)):*
<details><summary><i>Previously, sougou-bot wrote…</i></summary>
> ReverseMap is used to auto-populate a value if one was not provided. It's a rare use case. But if used, it will need to be efficient. The vindex will work even if you don't implement this function. So, if the efficient version is complicated, it's better to not implement it at all.
</details>
I pushed up the precomputed reverse map suggestion as it didn't seem like it would be much more complex, certainly better than what I had before.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1924)*

**Comment by chrisgillis** [comment]
Review status: 3 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 105 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOlBD_Srot4yjq0UkHV:-KOlBD_Srot4yjq0UkHW:b-f3k03e) ([raw file](https://github.com/youtube/vitess/blob/e7101e25f224ad3374af118c7e075bf4e008e020/go/vt/vtgate/vindexes/numeric_static_map.go#L105)):*
> ```Go
> 	id, ok := vind.reverseLookup[id]
> 	if !ok {
> 		return nil, fmt.Errorf("NumericStaticMap.ReverseMap: could not find key")
> ```

Should this return nil, error?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1924)*

**Comment by chrisgillis** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 74 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOhTiCJOtV9XQWE7M_R:-KOmfrnOUS8cdFGOtUP7:bs2rm8q) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L74)):*
<details><summary><i>Previously, enisoc-bot (Anthony Yeh) wrote…</i></summary>
> Here's an example of a lookup table that's not reversible:
> 
> ```json
> {
>   "10": 4
> }
> ```
> 
> This vindex as currently written would map 10->4, and since there is no entry for the key "4" in the lookup table, it would also map 4->4. You can't uniquely ReverseMap(4), because the answer is "either 4 or 10".
> 
> To avoid these overlaps, every rule in the lookup table must be part of some closed loop. For example, this table is reversible:
> 
> ```json
> {
>   "4": 10,
>   "10": 4
> }
> ```
> 
> So is this one:
> 
> ```json
> {
>   "4": 15,
>   "10": 4,
>   "15": 10
> }
> ```
> 
> So my question is, does your table generator use an algorithm that guarantees this will be the case?
> 
> As @sougou mentioned, it's not required that all vindexes be reversible. If your mapping is not reversible, you could just remove the `ReverseMap()` method so you don't claim to be.
> 
> But in general, it should be possible to make this kind of mapping scheme reversible, so it would be nice to make that guarantee since it can let VTGate infer some more things automatically.
</details>
Ah ha, thank you for spelling that out for me. I don't see any reason why I cannot generate the reverse maps as well in our generator. I think we can leave ReverseMap in for whatever advantages it provides vtgate.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 76 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOh69Lpd7oDbiffP8Jj:-KOmhK6Z02tP4_0hvK8s:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L76)):*
<details><summary><i>Previously, enisoc-bot (Anthony Yeh) wrote…</i></summary>
> It looks like you moved the one in `Verify()` but not the one here in `Map()`?
</details>
Done.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 105 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOlBD_Srot4yjq0UkHV:-KOmilnbtkbgK-ohfOZA:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/e7101e25f224ad3374af118c7e075bf4e008e020/go/vt/vtgate/vindexes/numeric_static_map.go#L105)):*
<details><summary><i>Previously, enisoc-bot (Anthony Yeh) wrote…</i></summary>
> `ReverseMap()` needs to also work for values that aren't in the lookup table - i.e. IDs that aren't in use yet at the time of your migration.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1924)*

**Comment by chrisgillis** [comment]
Review status: 3 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 74 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOhTiCJOtV9XQWE7M_R:-KOmoEsmhAwln2nJrPpw:b7z07wo) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L74)):*
<details><summary><i>Previously, chrisgillis (Chris Gillis) wrote…</i></summary>
> Ah ha, thank you for spelling that out for me. I don't see any reason why I cannot generate the reverse maps as well in our generator. I think we can leave ReverseMap in for whatever advantages it provides vtgate.
</details>
As I sat down to do this I started to see the complications. I will likely be removing reversemap for now. If we absolutely needed to support it we would probably need to jump the autoincrement on our shard key at the time of map generation to make enough room to support the reversal. I would like to investigate that, but not right now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1924)*

**Comment by chrisgillis** [comment]
Review status: 3 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[go/vt/vtgate/vindexes/numeric_static_map.go, line 74 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1924#-KOhTiCJOtV9XQWE7M_R:-KOmqvbvYNVY4S_tQc4S:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/e091e6220e7007340cda79c135b3989b7c4a40d8/go/vt/vtgate/vindexes/numeric_static_map.go#L74)):*
<details><summary><i>Previously, chrisgillis (Chris Gillis) wrote…</i></summary>
> As I sat down to do this I started to see the complications. I will likely be removing reversemap for now. If we absolutely needed to support it we would probably need to jump the autoincrement on our shard key at the time of map generation to make enough room to support the reversal. I would like to investigate that, but not right now.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1924)*

File: go/vt/vtgate/vindexes/numeric_static_map.go [hunk 1]
```diff
@@ -94,23 +98,6 @@ func (vind *NumericStaticMap) Map(_ VCursor, ids []interface{}) ([][]byte, error
...
-	if ok {
-		id = lookupNum
-	}
-	return id, nil
-}
-
-func init() {
-	Register("numeric_static_map", NewNumericStaticMap)
-}
-
```
Comment (chrisgillis) [inline go/vt/vtgate/vindexes/numeric_static_map.go]: Moved init() closer to top of file.

**Comment by chrisgillis** [comment]
Just wanted to double check to be absolutely sure, as per the discussion in the mailing list, this vindex and these changes will still allow us to split shards, right? 

Also as we are not implementing reversemap, is there any vtgate functionality we are losing?

**Comment by chrisgillis** [comment]
Thanks Sugu. Could someone retry the build on Travis? Not sure if reparenting is broken or if network issued caused the failure.

`2016/08/10 18:19:50 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp [::1]:17023: getsockopt: connection refused"; Reconnecting to {"localhost:17023" <nil>}
2016/08/10 18:19:50 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp [::1]:17023: getsockopt: connection refused"; Reconnecting to {"localhost:17023" <nil>}
-- 2016-08-10 18:19:50,623 utils:60 INFO ===== ======================================================================
-- 2016-08-10 18:19:50,623 utils:60 INFO ===== FAIL: test_reparent_during_worker_copy (__main__.TestReparentDuringWorkerCopy)
Simulates a destination reparent during a worker SplitClone copy.
-- 2016-08-10 18:19:50,623 utils:60 INFO ===== ----------------------------------------------------------------------
-- 2016-08-10 18:19:50,624 utils:60 INFO ===== Traceback (most recent call last):
  File "test/worker.py", line 546, in test_reparent_during_worker_copy
    self.verify_successful_worker_copy_with_reparent()
  File "test/worker.py", line 516, in verify_successful_worker_copy_with_reparent
    self.assertGreater(worker_vars['WorkerRetryCount'], 1)
AssertionError: 1 not greater than 1`

**Comment by chrisgillis** [comment]
Thanks, build is passing now. Anything left or can this be merged?

---

Review status: 2 of 4 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1924#-:-KOq-dPw1vMo9Cir84bP:b-8u7ucj)*

**Comment by enisoc** [comment]
LGTM

**Pull request merged**